src/tactic/positivity.lean

```diff
@@ -0,0 +1,360 @@
+        (λ _, failed),
+      pure (λ e, orelse' (base e) (t e)) },
+    dependencies := [] } }
+
+/-- Look for a proof of positivity/nonnegativity of an expression `e`; if found, return the proof
+together with a `strictness` stating whether the proof found was for strict positivity
+(`positive p`) or only for nonnegativity (`nonnegative p`). -/
+meta def core (e : expr) : tactic strictness := do
+  f ← attr.get_cache,
+  f e <|> fail "failed to prove positivity/nonnegativity"
+
+end positivity
+
+namespace interactive
+
+setup_tactic_parser
+
+/-- Tactic solving goals of the form `0 ≤ x` and `0 < x`.  The tactic works recursively according to
+the syntax of the expression `x`, if the atoms composing the expression all have numeric lower
+bounds which can be proved positive/nonnegative by `norm_num`.  This tactic either closes the goal
+or fails. -/
+meta def positivity : tactic unit := focus1 $ do
+  t ← target,
+  (rel_desired, a) ← match t with
+  | `(0 ≤ %%e₂) := pure (ff, e₂)
+  | `(%%e₂ ≥ 0) := pure (ff, e₂)
+  | `(0 < %%e₂) := pure (tt, e₂)
+  | `(%%e₂ > 0) := pure (tt, e₂)
+  | _ := fail "not a positivity/nonnegativity goal"
+  end,
+  strictness_proved ← tactic.positivity.core a,
+  match rel_desired, strictness_proved with
+  | tt, (positive p) := pure p
+  | tt, (nonnegative _) := fail ("failed to prove strict positivity, but it would be possible to" ++
+      "prove nonnegativity if desired")
+  | ff, (positive p) := mk_app ``le_of_lt [p]
+  | ff, (nonnegative p) := pure p
+  end >>= tactic.exact
+
+end interactive
+
+open positivity
+
+variables {R : Type*}
+
+/-! ### `positivity` extensions for particular arithmetic operations -/
+
+private lemma le_min_of_lt_of_le [linear_order R] (a b c : R) (ha : a < b) (hb : a ≤ c) :
+  a ≤ min b c :=
+le_min ha.le hb
+
+private lemma le_min_of_le_of_lt [linear_order R] (a b c : R) (ha : a ≤ b) (hb : a < c) :
+  a ≤ min b c :=
+le_min ha hb.le
+
+/-- Extension for the `positivity` tactic: the `min` of two numbers is nonnegative if both are
+nonnegative, and strictly positive if both are. -/
+@[positivity]
+meta def positivity_min : expr → tactic strictness
+| `(min %%a %%b) := do
+  strictness_a ← core a,
+  strictness_b ← core b,
+  match strictness_a, strictness_b with
+  | (positive pa), (positive pb) := positive <$> mk_app ``lt_min [pa, pb]
+  | (positive pa), (nonnegative pb) := nonnegative <$> mk_app ``le_min_of_lt_of_le [pa, pb]
+  | (nonnegative pa), (positive pb)  := nonnegative <$> mk_app ``le_min_of_le_of_lt [pa, pb]
+  | (nonnegative pa), (nonnegative pb)  := nonnegative <$> mk_app ``le_min [pa, pb]
+  end
+| _ := failed
+
+/-- Extension for the `positivity` tactic: the `max` of two numbers is nonnegative if at least one
+is nonnegative, and strictly positive if at least one is positive. -/
+@[positivity]
+meta def positivity_max : expr → tactic strictness
+| `(max %%a %%b) := tactic.positivity.orelse' (do
+      strictness_a ← core a,
+      match strictness_a with
```

Comment:
This pattern looks worthy of a `strictness.mmap` definition, given how often it appears

Reply:
That is, `mmap : strictness → (expr → tactic expr) → (expr → tactic expr) → tactic strictness

Reply:
It doesn't really generalize to multiple argument functions. I think it's better documenting as written